### ISSUE TITLE
HIP Memory Access Fault / GPU Reset for `stable-diffusion-burn`

### ISSUE THREAD

**Issue by Beinsezii** [tracel-ai/cubecl#457]
Context https://github.com/tracel-ai/burn/discussions/2709

Reproduction
```
git clone https://github.com/Beinsezii/bsz-sd-burn.git
cd bsz-sd-burn
wget https://huggingface.co/Gadersd/Stable-Diffusion-Burn/resolve/main/SDv1-4.mpk
cargo run --release --bin sample burn SDv1-4.mpk 7.5 20 "An ancient mossy stone." img
```

Using a 7900 XTX on ROCm 6.3.1 using the official ubuntu packages `https://repo.radeon.com/amdgpu/6.3.1/ubuntu` should trigger
`Memory access fault by GPU node-1 (Agent handle: 0x639280bf76c0) on address 0x71a5e8220000. Reason: Page not present or supervisor privilege.`
When trying to sample the model. Additionally at one point I had it trigger a GPU reset when adding debug printouts. Similar behavior was observed using the ROCm 6.2.4 Arch Linux packages.

Can provide more information on request.

**Comment by nathanielsimard** [comment]
I think the way we are doing check bounds in our matmul kernels isn't safe. It kind of work on CUDA, but could be improved.

```rust
let tmp1 = input[index]; // maybe out of bound
let value = select(index < input.buffer_len(), tmp1, 0u32);
```

**Comment by Beinsezii** [comment]
The fun part is CUDA also has ways to explode with an incorrect bounds check.
```python
import torch

device = torch.device("cuda")

embedding = torch.nn.Embedding(num_embeddings=10, embedding_dim=3).to(device)
input_data = torch.tensor([0, 1, 10], device=device, dtype=torch.long)

try:
    output = embedding(input_data)
except:
    pass

unrelated = torch.tensor([1, 2, 3], device=device, dtype=torch.float32)
print("Banana")
```
NVIDIA users can't print `banana` because a device side assert when indexing the embedding poisons the entire CUDA context.

**Comment by Beinsezii** [comment]
Seems like its on

`burn_tensor::tensor::activation::base::softmax(qk, 3)`

where `qk` has shape `[1, 12, 2, 2]`

**Comment by Beinsezii** [comment]
I believe this will reproduce it

```rust
    type Backend = burn::backend::hip_jit::Hip;
    let device = burn::backend::hip_jit::HipDevice::new(0);

    let data: Vec<u8> = vec![
        127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255,
        255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255,
        127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255,
        255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255,
        127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255,
        255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255,
        127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255,
        255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255,
        127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255,
        255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255,
        127, 255, 255, 255, 127, 255, 255, 255, 127, 255, 255, 255,
    ];

    println!("td init");
    let td = TensorData::from_bytes(data, [1, 12, 2, 2], burn::tensor::DType::F32);
    println!("t init");
    let t: Tensor<Backend, 4> = Tensor::from_data(td, &device);
    println!("softmax");
    let sm = burn::tensor::activation::softmax(t, 3);
    println!("done");
```

As it can trigger a memory access fault immediately after `println!("softmax");` before anything else in the program runs

**Comment by Beinsezii** [comment]
The matmul NaN can be reproduced like

```rust
fn main() {
    type Backend = burn::backend::hip_jit::Hip;
    let device = burn::backend::hip_jit::HipDevice::new(0);

    let dist = burn::tensor::Distribution::Uniform(-1.0, 1.0);
    let v = burn::tensor::Tensor::<Backend, 4>::random([1, 12, 2, 64], dist, &device);
    let w = burn::tensor::Tensor::<Backend, 4>::random([1, 12, 2, 2], dist, &device);
    let o = w.matmul(v);
    assert!(!o.is_nan().any().into_scalar());
}
```

vs PT
```python
import torch

v = torch.randn([1, 12, 2, 64], dtype=torch.float32, device="cuda")
w = torch.randn([1, 12, 2, 2], dtype=torch.float32, device="cuda")
assert w.matmul(v).isfinite().all().item()
```

**Comment by syl20bnr** [comment]
@Beinsezii do you have stability improvements with the lastest `0.17` version of Burn ?

**Comment by Beinsezii** [comment]
It ate > 20GiB of memory on the decode then OOM'd so I'm guessing some code needs to change somewhere now, but the matmul and softmax no longer break including w/ the minimal examples.
